File: drivers/syscall_fs.c
#include "syscall_common.h"
#include "../fs/fileio.h"
#include "../fs/ramfs.h"
#include "../lib/user_syscall_defs.h"
#include "../mm/user_copy.h"
#include "../mm/user_copy_helpers.h"
#include "../lib/string.h"
#include "../lib/klog.h"
#include "../mm/mm_constants.h"
#include "../mm/kernel_heap.h"

#define USER_FS_MAX_ENTRIES 64

static enum syscall_disposition syscall_fs_error(struct interrupt_frame *frame) {
    return syscall_return_err(frame, -1);
}

enum syscall_disposition syscall_fs_open(task_t *task, struct interrupt_frame *frame) {
    if (!task || task->process_id == INVALID_PROCESS_ID) {
        return syscall_fs_error(frame);
    }
    char path[USER_PATH_MAX];
    if (syscall_copy_user_str(path, sizeof(path), (const char *)frame->rdi) != 0) {
        return syscall_fs_error(frame);
    }

    uint32_t flags = (uint32_t)frame->rsi;
    int fd = file_open_for_process(task->process_id, path, flags);
    if (fd < 0) {
        return syscall_fs_error(frame);
    }
    return syscall_return_ok(frame, (uint64_t)fd);
}

enum syscall_disposition syscall_fs_close(task_t *task, struct interrupt_frame *frame) {
    if (!task || task->process_id == INVALID_PROCESS_ID) {
        return syscall_fs_error(frame);
    }
    if (file_close_fd(task->process_id, (int)frame->rdi) != 0) {
        return syscall_fs_error(frame);
    }
    return syscall_return_ok(frame, 0);
}

enum syscall_disposition syscall_fs_read(task_t *task, struct interrupt_frame *frame) {
    if (!task || task->process_id == INVALID_PROCESS_ID || !frame->rsi) {
        return syscall_fs_error(frame);
    }
    char tmp[USER_IO_MAX_BYTES];
    size_t request_len = frame->rdx > USER_IO_MAX_BYTES ? USER_IO_MAX_BYTES : (size_t)frame->rdx;
    ssize_t bytes = file_read_fd(task->process_id, (int)frame->rdi, tmp, request_len);
    if (bytes < 0) {
        return syscall_fs_error(frame);
    }
    if (syscall_copy_to_user_bounded((void *)frame->rsi, tmp, (size_t)bytes) != 0) {
        return syscall_fs_error(frame);
    }
    return syscall_return_ok(frame, (uint64_t)bytes);
}

enum syscall_disposition syscall_fs_write(task_t *task, struct interrupt_frame *frame) {
    if (!task || task->process_id == INVALID_PROCESS_ID || !frame->rsi) {
        return syscall_fs_error(frame);
    }
    char tmp[USER_IO_MAX_BYTES];
    size_t write_len = 0;
    if (syscall_bounded_from_user(tmp, sizeof(tmp), (const void *)frame->rsi,
                                  frame->rdx, USER_IO_MAX_BYTES, &write_len) != 0) {
        return syscall_fs_error(frame);
    }

    ssize_t bytes = file_write_fd(task->process_id, (int)frame->rdi, tmp, write_len);
    if (bytes < 0) {
        return syscall_fs_error(frame);
    }
    return syscall_return_ok(frame, (uint64_t)bytes);
}

enum syscall_disposition syscall_fs_stat(task_t *task, struct interrupt_frame *frame) {
    if (!task || !frame->rdi || !frame->rsi) {
        return syscall_fs_error(frame);
    }
    char path[USER_PATH_MAX];
    if (syscall_copy_user_str(path, sizeof(path), (const char *)frame->rdi) != 0) {
        return syscall_fs_error(frame);
    }

    ramfs_node_t *node = ramfs_acquire_node(path);
    if (!node) {
        return syscall_fs_error(frame);
    }

    user_fs_stat_t stat = {0};
    stat.size = (uint32_t)ramfs_get_size(node);
    stat.type = (node->type == RAMFS_TYPE_DIRECTORY) ? 1 : (node->type == RAMFS_TYPE_FILE ? 0 : 0xFF);
    ramfs_node_release(node);

    if (syscall_copy_to_user_bounded((void *)frame->rsi, &stat, sizeof(stat)) != 0) {
        return syscall_fs_error(frame);
    }

    return syscall_return_ok(frame, 0);
}

enum syscall_disposition syscall_fs_mkdir(task_t *task, struct interrupt_frame *frame) {
    (void)task;
    char path[USER_PATH_MAX];
    if (syscall_copy_user_str(path, sizeof(path), (const char *)frame->rdi) != 0) {
        return syscall_fs_error(frame);
    }

    if (!ramfs_create_directory(path)) {
        return syscall_fs_error(frame);
    }
    return syscall_return_ok(frame, 0);
}

enum syscall_disposition syscall_fs_unlink(task_t *task, struct interrupt_frame *frame) {
    (void)task;
    char path[USER_PATH_MAX];
    if (syscall_copy_user_str(path, sizeof(path), (const char *)frame->rdi) != 0) {
        return syscall_fs_error(frame);
    }

    if (file_unlink_path(path) != 0) {
        return syscall_fs_error(frame);
    }
    return syscall_return_ok(frame, 0);
}

enum syscall_disposition syscall_fs_list(task_t *task, struct interrupt_frame *frame) {
    (void)task;
    char path[USER_PATH_MAX];
    if (syscall_copy_user_str(path, sizeof(path), (const char *)frame->rdi) != 0 || frame->rsi == 0) {
        return syscall_fs_error(frame);
    }

    user_fs_list_t list_hdr;
    if (user_copy_from_user(&list_hdr, (const void *)frame->rsi, sizeof(list_hdr)) != 0) {
        return syscall_fs_error(frame);
    }

    uint32_t cap = list_hdr.max_entries;
    if (cap == 0 || cap > USER_FS_MAX_ENTRIES || !list_hdr.entries) {
        return syscall_fs_error(frame);
    }

    ramfs_node_t **entries = NULL;
    int count = 0;
    if (ramfs_list_directory(path, &entries, &count) != 0) {
        return syscall_fs_error(frame);
    }

    if (count < 0) count = 0;
    if ((uint32_t)count > cap) count = (int)cap;

    user_fs_entry_t *tmp = (user_fs_entry_t *)kmalloc(sizeof(user_fs_entry_t) * (size_t)cap);
    if (!tmp) {
        if (entries) {
            ramfs_release_list(entries, count);
            kfree(entries);
        }
        return syscall_fs_error(frame);
    }

    for (int i = 0; i < count; i++) {
        ramfs_node_t *e = entries[i];
        if (!e) {
            tmp[i].name[0] = '\0';
            tmp[i].type = 0;
            tmp[i].size = 0;
            continue;
        }
        size_t nlen = strlen(e->name);
        if (nlen >= sizeof(tmp[i].name)) {
            nlen = sizeof(tmp[i].name) - 1;
        }
        for (size_t j = 0; j < nlen; j++) {
            tmp[i].name[j] = e->name[j];
        }
        tmp[i].name[nlen] = '\0';
        tmp[i].type = (e->type == RAMFS_TYPE_DIRECTORY) ? 1 : 0;
        tmp[i].size = (uint32_t)e->size;
    }

    list_hdr.count = (uint32_t)count;

    int rc = user_copy_to_user(list_hdr.entries, tmp, sizeof(user_fs_entry_t) * (size_t)count);
    if (rc == 0) {
        rc = user_copy_to_user((void *)frame->rsi, &list_hdr, sizeof(list_hdr));
    }

    if (entries) {
        ramfs_release_list(entries, count);
        kfree(entries);
    }
    kfree(tmp);

    if (rc != 0) {
        return syscall_fs_error(frame);
    }

    return syscall_return_ok(frame, 0);
}

File: drivers/src/syscall_fs.rs
//! Filesystem syscall handlers
//!
//! All handlers use SyscallContext for safe pointer access.

#![allow(clippy::too_many_arguments)]

use core::ffi::{CStr, c_char, c_int, c_void};
use core::{mem, ptr};

use crate::syscall_common::{
    SyscallDisposition, USER_IO_MAX_BYTES, USER_PATH_MAX, syscall_bounded_from_user,
    syscall_copy_to_user_bounded, syscall_copy_user_str, syscall_return_err,
};
use crate::syscall_context::SyscallContext;
use crate::syscall_types::{INVALID_PROCESS_ID, InterruptFrame, Task};

const USER_FS_MAX_ENTRIES: u32 = 64;
type RamfsNode = ramfs_node_t;

#[repr(C)]
pub struct UserFsEntry {
    pub name: [c_char; 64],
    pub type_: u8,
    pub size: u32,
}

#[repr(C)]
pub struct UserFsStat {
    pub type_: u8,
    pub size: u32,
}

#[repr(C)]
pub struct UserFsList {
    pub entries: *mut UserFsEntry,
    pub max_entries: u32,
    pub count: u32,
}

use slopos_fs::fileio::{
    file_close_fd, file_open_for_process, file_read_fd, file_unlink_path, file_write_fd,
};
use slopos_fs::ramfs::{
    RAMFS_TYPE_DIRECTORY, RAMFS_TYPE_FILE, ramfs_acquire_node, ramfs_create_directory,
    ramfs_get_size, ramfs_list_directory, ramfs_node_release, ramfs_node_t, ramfs_release_list,
};

use slopos_mm::kernel_heap::{kfree, kmalloc};
use slopos_mm::user_copy::{user_copy_from_user, user_copy_to_user};

pub fn syscall_fs_open(task: *mut Task, frame: *mut InterruptFrame) -> SyscallDisposition {
    let Some(ctx) = SyscallContext::new(task, frame) else {
        return syscall_return_err(frame, u64::MAX);
    };

    let pid = match ctx.process_id() {
        Some(p) if p != INVALID_PROCESS_ID => p,
        _ => return ctx.err(),
    };

    let args = ctx.args();
    let mut path = [0i8; USER_PATH_MAX];
    if syscall_copy_user_str(path.as_mut_ptr(), path.len(), args.arg0_const_ptr()) != 0 {
        return ctx.err();
    }

    let flags = args.arg1_u32();
    let fd = file_open_for_process(pid, path.as_ptr(), flags);
    if fd < 0 {
        ctx.err()
    } else {
        ctx.ok(fd as u64)
    }
}

pub fn syscall_fs_close(task: *mut Task, frame: *mut InterruptFrame) -> SyscallDisposition {
    let Some(ctx) = SyscallContext::new(task, frame) else {
        return syscall_return_err(frame, u64::MAX);
    };

    let pid = match ctx.process_id() {
        Some(p) if p != INVALID_PROCESS_ID => p,
        _ => return ctx.err(),
    };

    let args = ctx.args();
    let fd = args.arg0 as c_int;
    let rc = file_close_fd(pid, fd);
    if rc != 0 {
        ctx.err()
    } else {
        ctx.ok(0)
    }
}

pub fn syscall_fs_read(task: *mut Task, frame: *mut InterruptFrame) -> SyscallDisposition {
    let Some(ctx) = SyscallContext::new(task, frame) else {
        return syscall_return_err(frame, u64::MAX);
    };

    let pid = match ctx.process_id() {
        Some(p) if p != INVALID_PROCESS_ID => p,
        _ => return ctx.err(),
    };

    let args = ctx.args();
    if args.arg1 == 0 {
        return ctx.err();
    }

    let mut tmp = [0u8; USER_IO_MAX_BYTES];
    let request_len = args.arg2_usize();
    let capped_len = request_len.min(USER_IO_MAX_BYTES);

    let bytes = file_read_fd(
        pid,
        args.arg0 as c_int,
        tmp.as_mut_ptr() as *mut c_char,
        capped_len,
    );
    if bytes < 0 {
        return ctx.err();
    }

    let copy_len = bytes as usize;
    if syscall_copy_to_user_bounded(
        args.arg1_ptr(),
        tmp.as_ptr() as *const c_void,
        copy_len,
    ) != 0
    {
        return ctx.err();
    }

    ctx.ok(bytes as u64)
}

pub fn syscall_fs_write(task: *mut Task, frame: *mut InterruptFrame) -> SyscallDisposition {
    let Some(ctx) = SyscallContext::new(task, frame) else {
        return syscall_return_err(frame, u64::MAX);
    };

    let pid = match ctx.process_id() {
        Some(p) if p != INVALID_PROCESS_ID => p,
        _ => return ctx.err(),
    };

    let args = ctx.args();
    if args.arg1 == 0 {
        return ctx.err();
    }

    let mut tmp = [0u8; USER_IO_MAX_BYTES];
    let mut write_len: usize = 0;
    if syscall_bounded_from_user(
        tmp.as_mut_ptr() as *mut c_void,
        tmp.len(),
        args.arg1 as *const c_void,
        args.arg2,
        USER_IO_MAX_BYTES,
        &mut write_len as *mut usize,
    ) != 0
    {
        return ctx.err();
    }

    let bytes = file_write_fd(
        pid,
        args.arg0 as c_int,
        tmp.as_ptr() as *const c_char,
        write_len,
    );
    if bytes < 0 {
        ctx.err()
    } else {
        ctx.ok(bytes as u64)
    }
}

pub fn syscall_fs_stat(task: *mut Task, frame: *mut InterruptFrame) -> SyscallDisposition {
    let Some(ctx) = SyscallContext::new(task, frame) else {
        return syscall_return_err(frame, u64::MAX);
    };

    let args = ctx.args();
    if args.arg0 == 0 || args.arg1 == 0 {
        return ctx.err();
    }

    let mut path = [0i8; USER_PATH_MAX];
    if syscall_copy_user_str(path.as_mut_ptr(), path.len(), args.arg0_const_ptr()) != 0 {
        return ctx.err();
    }

    let node = ramfs_acquire_node(path.as_ptr());
    if node.is_null() {
        return ctx.err();
    }

    let mut stat = UserFsStat { type_: 0, size: 0 };
    unsafe {
        stat.size = ramfs_get_size(node) as u32;
        let kind = (*node).type_;
        stat.type_ = if kind == RAMFS_TYPE_DIRECTORY {
            1
        } else if kind == RAMFS_TYPE_FILE {
            0
        } else {
            0xFF
        };
        ramfs_node_release(node);
    }

    if syscall_copy_to_user_bounded(
        args.arg1_ptr(),
        &stat as *const _ as *const c_void,
        mem::size_of::<UserFsStat>(),
    ) != 0
    {
        return ctx.err();
    }

    ctx.ok(0)
}

pub fn syscall_fs_mkdir(task: *mut Task, frame: *mut InterruptFrame) -> SyscallDisposition {
    let Some(ctx) = SyscallContext::new(task, frame) else {
        return syscall_return_err(frame, u64::MAX);
    };

    let args = ctx.args();
    let mut path = [0i8; USER_PATH_MAX];
    if syscall_copy_user_str(path.as_mut_ptr(), path.len(), args.arg0_const_ptr()) != 0 {
        return ctx.err();
    }

    let created = ramfs_create_directory(path.as_ptr());
    if !created.is_null() {
        ctx.ok(0)
    } else {
        ctx.err()
    }
}

pub fn syscall_fs_unlink(task: *mut Task, frame: *mut InterruptFrame) -> SyscallDisposition {
    let Some(ctx) = SyscallContext::new(task, frame) else {
        return syscall_return_err(frame, u64::MAX);
    };

    let args = ctx.args();
    let mut path = [0i8; USER_PATH_MAX];
    if syscall_copy_user_str(path.as_mut_ptr(), path.len(), args.arg0_const_ptr()) != 0 {
        return ctx.err();
    }

    if file_unlink_path(path.as_ptr()) != 0 {
        ctx.err()
    } else {
        ctx.ok(0)
    }
}

pub fn syscall_fs_list(task: *mut Task, frame: *mut InterruptFrame) -> SyscallDisposition {
    let Some(ctx) = SyscallContext::new(task, frame) else {
        return syscall_return_err(frame, u64::MAX);
    };

    let args = ctx.args();
    let mut path = [0i8; USER_PATH_MAX];

    if syscall_copy_user_str(path.as_mut_ptr(), path.len(), args.arg0_const_ptr()) != 0
        || args.arg1 == 0
    {
        return ctx.err();
    }

    let mut list_hdr = UserFsList {
        entries: ptr::null_mut(),
        max_entries: 0,
        count: 0,
    };

    if user_copy_from_user(
        &mut list_hdr as *mut _ as *mut c_void,
        args.arg1 as *const c_void,
        mem::size_of::<UserFsList>(),
    ) != 0
    {
        return ctx.err();
    }

    let cap = list_hdr.max_entries;
    if cap == 0 || cap > USER_FS_MAX_ENTRIES || list_hdr.entries.is_null() {
        return ctx.err();
    }

    let mut entries: *mut *mut RamfsNode = ptr::null_mut();
    let mut count: c_int = 0;
    let rc = ramfs_list_directory(path.as_ptr(), &mut entries, &mut count);
    if rc != 0 {
        return ctx.err();
    }

    if count < 0 {
        count = 0;
    }
    if (count as u32) > cap {
        count = cap as c_int;
    }

    let tmp_size = mem::size_of::<UserFsEntry>() * cap as usize;
    let tmp_ptr = kmalloc(tmp_size) as *mut UserFsEntry;
    if tmp_ptr.is_null() {
        if !entries.is_null() {
            ramfs_release_list(entries, count);
            kfree(entries as *mut c_void);
        }
        return ctx.err();
    }
    unsafe {
        core::ptr::write_bytes(tmp_ptr as *mut u8, 0, tmp_size);
    }

    for i in 0..(count as usize) {
        unsafe {
            let entry_ptr = *entries.add(i);
            let dst = &mut *tmp_ptr.add(i);
            if entry_ptr.is_null() {
                dst.type_ = 0;
                dst.size = 0;
                continue;
            }

            let cstr = CStr::from_ptr((*entry_ptr).name);
            let name_bytes = cstr.to_bytes();
            let nlen = name_bytes.len().min(dst.name.len());
            for (dst_byte, src_byte) in dst.name[..nlen].iter_mut().zip(name_bytes[..nlen].iter()) {
                *dst_byte = *src_byte as i8;
            }
            if nlen < dst.name.len() {
                dst.name[nlen] = 0;
            }
            dst.type_ = if (*entry_ptr).type_ == RAMFS_TYPE_DIRECTORY {
                1
            } else {
                0
            };
            dst.size = (*entry_ptr).size as u32;
        }
    }

    list_hdr.count = count as u32;

    let mut rc_user = user_copy_to_user(
        list_hdr.entries as *mut c_void,
        tmp_ptr as *const c_void,
        mem::size_of::<UserFsEntry>() * count as usize,
    );
    if rc_user == 0 {
        rc_user = user_copy_to_user(
            args.arg1 as *mut c_void,
            &list_hdr as *const _ as *const c_void,
            mem::size_of::<UserFsList>(),
        );
    }

    if !entries.is_null() {
        ramfs_release_list(entries, count);
        kfree(entries as *mut c_void);
    }
    kfree(tmp_ptr as *mut c_void);

    if rc_user != 0 {
        ctx.err()
    } else {
        ctx.ok(0)
    }
}
